File: web/src/main/java/no/nav/sbl/dialogarena/mininnboks/consumer/domain/Henvendelse.java
package no.nav.sbl.dialogarena.mininnboks.consumer.domain;

import org.joda.time.DateTime;

import java.io.Serializable;
import java.util.Comparator;

import static java.util.Collections.reverseOrder;

public class Henvendelse implements Serializable {

    public String id;
    public String traadId, fritekst, kanal, eksternAktor, brukersEnhet, tilknyttetEnhet, temagruppeNavn, statusTekst, kontorsperreEnhet, temaNavn, korrelasjonsId;
    public Henvendelsetype type;
    public Temagruppe temagruppe;
    public DateTime opprettet, avsluttet;
    public Boolean fraNav, fraBruker, kassert = false, erTilknyttetAnsatt;
    private DateTime lestDato;

    public Henvendelse(String id) {
        this.id = id;
    }

    public Henvendelse(String fritekst, Temagruppe temagruppe) {
        this.fritekst = fritekst;
        this.temagruppe = temagruppe;
    }

    public Henvendelse withTraadId(String traadId) {
        this.traadId = traadId;
        return this;
    }

    public Henvendelse withType(Henvendelsetype type) {
        this.type = type;
        return this;
    }

    public Henvendelse withOpprettetTid(DateTime opprettetTid) {
        this.opprettet = opprettetTid;
        return this;
    }

    public void markerSomLest(DateTime lestDato) {
        this.lestDato = lestDato;
    }

    public void markerSomLest() {
        this.lestDato = DateTime.now();
    }

    public DateTime getLestDato() {
        return lestDato;
    }

    public boolean isLest() {
        return lestDato != null;
    }

    public static final Comparator<Henvendelse> NYESTE_OVERST = reverseOrder(Comparator.comparing(henvendelse -> henvendelse.opprettet));

    public Henvendelse withTemaNavn(String temaNavn) {
        this.temaNavn = temaNavn;
        return this;
    }

}

File: src/main/kotlin/no/nav/sbl/dialogarena/mininnboks/consumer/domain/Henvendelse.kt
package no.nav.sbl.dialogarena.mininnboks.consumer.domain

import java.util.*
import kotlin.collections.ArrayList

data class Henvendelse(val id: String? = null,
                       val traadId: String? = null,
                       val fritekst: String? = null,
                       val kanal: String? = null,
                       val eksternAktor: String? = null,
                       val brukersEnhet: String? = null,
                       val tilknyttetEnhet: String? = null,
                       val temagruppeNavn: String? = null,
                       val statusTekst: String? = null,
                       val kontorsperreEnhet: String? = null,
                       val temaNavn: String? = null,
                       val temaKode: String? = null,
                       val korrelasjonsId: String? = null,
                       val journalpostId: String? = null,
                       val dokumentIdListe: ArrayList<String> = ArrayList(),
                       val oppgaveType: String? = null,
                       val oppgaveUrl: String? = null,
                       val type: Henvendelsetype,
                       val temagruppe: Temagruppe? = null,
                       val opprettet: Date? = null,
                       val avsluttet: Date? = null,
                       val fraNav: Boolean? = null,
                       val fraBruker: Boolean? = null,
                       val kassert: Boolean = false,
                       val erTilknyttetAnsatt: Boolean? = null,
                       val ferdigstiltUtenSvar: Boolean? = null,
                       val lestDato: Date? = null) {

    val isLest: Boolean
        get() = lestDato != null

    companion object {
        val NYESTE_OVERST: Comparator<Henvendelse?> = Collections.reverseOrder(Comparator.comparing { henvendelse: Henvendelse -> henvendelse.opprettet })
    }
}
